src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
         Ok(NodeMap { node_map: out_map })
     }
 
+    /// Substitute a set of nodes with a single node.
+    ///
+    /// :param set to_replace: A set of nodes to be removed and replaced
+    ///     by the specified node.
+    /// :param int node: An existing node to use as the replacement.
+    /// :param bool force_check_cycle: If set to ``True``, validates
+    ///     that the substitution will not introduce cycles before
+    ///     modifying the graph, even if cycle check is disabled for
+    ///     this instance of :class:`retworkx.PyDiGraph`. Otherwise,
+    ///     the value of member ``check_cycle`` is inherited.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    /// :raises ValueError: ``to_replace`` is empty or ``node``
+    ///     has existing connections.
+    #[pyo3(
+        text_signature = "(self, to_replace, node, /, force_check_cycle=None)"
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
```

Comment:
You can replace this with a single iterator chain like:

```suggestion
            let mut visit_next: VecDeque<NodeIndex> = nodes
                .iter()
                .map(|n| graph.edges(*n))
                .flatten()
                .filter_map(|edge| {
                    let target_node = edge.target();
                    if !nodes.contains(&target_node) {
                        Some(target_node)
                    } else {
                        None
                    }
                })
                .collect();
```

This has the advantage of allocating at creation time so it should be slightly more efficient

Reply:
Beautiful! TIL rustc has very similar collection APIs to Scala.

Reply:
I was able to make this even _slightly_ nicer by using `flat_map` instead of `map` followed by `flatten` :).